DX: k3s/k3d instead of kind & CI: autohttps testing (#1664)
jupyterhub/zero-to-jupyterhub-k8s: CONTRIBUTING.md

- [helm](https://helm.sh/docs/intro/install/) (also [setup tab completion](https://helm.sh/docs/helm/helm_completion/))
- Python 3.7+ ([Anaconda.com](https://www.anaconda.com/distribution/),
  [Python.org](https://www.python.org/downloads/)) and dependencies:
  ```shell
  pip install -r dev-requirements.txt
  pip install -r doc/doc-requirements.txt
  ```

You can check if you have it all like this:

```shell
docker --version
kubectl version --client
helm version
pytest --version
chartpress --version
```

## 2: Setup a Kubernetes cluster

We need a Kubernetes cluster to work against.

### Linux only: Kubernetes setup with k3s
With [k3s](https://github.com/rancher/k3s) we can _quickly_ create a Kubernetes
cluster, and we _don't have to transfer docker images_ built on our computer to
make them available in the Kubernetes cluster.

__Install__

```shell
# Installs a ~50 MB k3s binary, setups and starts a systemctl service called
# k3s which is also enabled to run on startup, provides a k3s-uninstall.sh

> # script, disables not needed functionality. You will be asked for sudo rights

Author: I'd be fine with that potential issue, since k3s is in the CI system + I'd use it for local dev so I feel confident we would catch issues quite quick.

curl -sfL https://get.k3s.io | sh -s - \
    --write-kubeconfig-mode=644 \
    --disable metrics-server \
    --disable traefik \
    --disable local-storage \

>     --disable-network-policy \

Author: Oh... I forgot that now we may want to have the step to install the CNI in this docs as well, because --flannel-backend=none is only suitable along with setting up calico.

>     --disable-network-policy \

Author: Hmmmm... I wonder if i think it make more sense to have a section on running netpol tests. They add quite a bit of complexity to the setup, needing --flannel-backend=none during k3s install + calico installed afterwards with a patched configuration. And, i've not tested it on k3d either and bet it can be different in that situation.

WDYT, okay to make a section about netpol tests that we reference instead of forcing everybody to install the tools needed for them?

>     --disable-network-policy \

Author: I'll make another pass on the netpol instructions, perhaps clearing away canal steps that didn't pan out in Travis CI which reduce some complexity and avoids raising questions about canal etc.

I'm strongly biased towards:
- only supporting what we are confident we can maintain
- minimal amount of helper scripts

# Contributing

Welcome dear open source colleague! As this is a [Jupyter](https://jupyter.org)
project please start by looking at the [Jupyter contributor
guide](https://jupyter.readthedocs.io/en/latest/contributor/content-contributor.html),
and follow [Project Jupyter's Code of
Conduct](https://github.com/jupyter/governance/blob/master/conduct/code_of_conduct.md)
to help us sustain a warm and welcoming collaborative environment.

If you don't have [git](https://www.git-scm.com/) already, install it and clone
this repository.

```shell
git clone https://github.com/jupyterhub/zero-to-jupyterhub-k8s
```

# Setting up for documentation development

See [doc/README.md](doc/README.md).

# Setting up for Helm chart development

## 1: Prerequisites

This needs to be installed:

- [docker](https://docs.docker.com/get-docker/)
- [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/) (also [setup tab completion](https://kubernetes.io/docs/tasks/tools/install-kubectl/#enabling-shell-autocompletion))
- [helm](https://helm.sh/docs/intro/install/) (also [setup tab completion](https://helm.sh/docs/helm/helm_completion/))

> - Python 3.7+ ([Anaconda.com](https://www.anaconda.com/distribution/),

Author: Hmmm... I dont remember if this is needed, but for a local dev setup, id be fine asking for py37+. I remember some differences of relevance with subprocess.popen etc that made me like to not need to support py36.

I bumped the ci system to 3.7, and this reflects that only i figure, that we dont test on an older version only impacts the local dev scripts i guess

> - Python 3.7+ ([Anaconda.com](https://www.anaconda.com/distribution/),

Reviewer: I've tested this with Python 3.6 due to using a Ubuntu 18.04 VM (see my other comment). Chartpress officially supports 3.6+, so can we change this to 3.6 to make it easier for others to setup their environment? Requiring Anaconda is an extra step.

# 30443: proxy-public - where you reach both /hub/home and /user/myuser (HTTPS)
# 32444: pebble - where you reach /roots/0 on Pebble's management API (HTTPS)
k3d create --enable-registry --wait 60 --publish 30443:30443 --publish 32444:32444
export KUBECONFIG="$(k3d get-kubeconfig --name='k3s-default')"
```

## 3: Install a local ACME server

Testing automatic TLS certificate acquisition with an ACME server like Let's
Encrypt from a local Kubernetes cluster is tricky. First you need a public
domain name registered and pointing to some public IP, and you need traffic to
that IP end up inside your Kubernetes cluster. In our TravisCI setup we must
install a local ACME server instead, and that is also recommended for local
development.

Pebble is a an ACME server like Let's Encrypt solely meant for testing purposes.
For more information, see
[jupyterhub/pebble-helm-chart](https://github.com/jupyterhub/pebble-helm-chart).

__Install Pebble__

```shell
helm repo add jupyterhub https://jupyterhub.github.io/helm-chart/
helm repo update
helm install pebble jupyterhub/pebble --values dev-config.yaml
```

## 4: Build images, update values, install chart

> This repository contains various `Dockerfile`s that builds to docker images in
> use by the Helm chart. To help us build these docker images only when needed,

Reviewer: ```suggestion
This repository contains various `Dockerfile`s that are used to build docker images
required by the Helm chart. To help us build these docker images only when needed,
```

and to update the Helm chart's [values.yaml](jupyterhub/values.yaml) file to use
the most recent image, we rely on a command line tool called
[`chartpress`](https://github.com/jupyterhub/chartpress) that is installed as
part of [dev-requirements.txt](dev-requirements.txt).

Chartpress is configured through [chartpress.yaml](chartpress.yaml), and will
only rebuild images if their dependent files in their respective directories or
`chartpress.yaml` itself has changed.

> 1. Ensure the latest git tags are available locally, as `chartpress` use them.

Reviewer: ```suggestion
1. Ensure the latest git tags are available locally, as `chartpress` uses them.
```

   ```shell
   git fetch origin --tags
   ```

1. Use [`chartpress`](https://github.com/jupyterhub/chartpress) to rebuild

>    images that needs to be rebuilt and update the chart's

Reviewer: ```suggestion
   images that need to be rebuilt and update the chart's
```

   [values.yaml](jupyterhub/values.yaml) file with the appropriate image tags.

   ```shell
   # run this if you are using k3s, or generally when your Kubernetes cluster
   # can make direct use of your locally built images
   chartpress

   # run this if you are using k3d

>    chartpress --image-prefix=registry.local:5000/ --push

Author: Lets try it! I figured it could cause issues as localhost could refer to other things in a kubernetes cluster than the actual computer with the k8s cluster.

>    chartpress --image-prefix=registry.local:5000/ --push

Author: If that is an issue, we use local.jovyan.org instead.

>    chartpress --image-prefix=registry.local:5000/ --push

Author: Thanks! I needed to use local.jovyan.org for it to work all the way, but it worked. The PR to fix it is https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/1696.

   ```

1. Use `helm` to upgrade (and install) your Helm chart.

   ```shell
   # note that --cleanup-on-fail is a very good practice to avoid
   # "<resource name> already exist" in future upgrades following

>    # a failed upgrade.

Reviewer: To me this sort of informative comment would be easier to read as part of the text above, instead of in the code block.

   helm upgrade --install jupyterhub ./jupyterhub --cleanup-on-fail --values dev-config.yaml
   ```

## 5: Visit the JupyterHub

> Soon you should be able to visit https://local.jovyan.org:30443 where you
> manually need to choose to accept to trust certificates signed by a certificate
> authority you don't already trust.

Reviewer: ```suggestion
Soon you will be able to visit https://local.jovyan.org:30443 where you
must accept the certificates signed by the Pebble certificate authority used for testing.
```

## 6: Run tests

The test suite runs outside your Kubernetes cluster.

```shell
pytest -vx -m "not netpol" ./tests
```

Note that we have disabled the creation of NetworkPolicy Kubernetes resources in

> our [dev-config.yaml](dev-config.yaml) and we choose to not run a test of them

Reviewer: I thought you fixed this by installing calico/canal?

Author: Outdated remnant, good catch! Calico btw, i failed after lots of attempt with k3s in docker mode + canal on travis

at the moment. This is because the built in NetworkPolicy controller in k3s is
not robustly enforcing them, as discussed in [this
issue](https://github.com/rancher/k3s/issues/947#issuecomment-627641541).

# Debugging

Various things can go wrong while working with the local development
environment, here are some typical issues and what to do about them.

## Basic debugging strategy in Kubernetes

A good debugging strategy is to start with the following steps.

1. Inspect the status of pods with `kubectl get pods`.
2. Inspect events and status of some pod with `kubectl describe pod <name>`.
3. Inspect a pod's container's logs with `kubectl logs ...`. Sometimes you need
   to specify `-c <container name>` or `--all-containers`. And sometimes you may
   want to specify the `--previous` flag to see the logs from the previous
   container run.

## HTTPS errors

> If you experience the need to confirm that you trust the certificate, that is

Reviewer: ```suggestion
If you are asked to trust the certificate, that is
```

expected, as the certificate authority isn't and shouldn't be trusted unless it
is solely for testing purposes.

But if Chrome presents `ERR_SSL_PROTOCOL_ERROR` or Firefox presents

> `SSL_ERROR_INTERNARROR_ALERT`, then the autohttps pod have probably failed to

Reviewer: ```suggestion
`SSL_ERROR_INTERNARROR_ALERT`, then the `autohttps` pod has probably failed to
```

acquire a TLS certificate from the ACME server.

```shell
# the certificate should be available here
kubectl exec -it deploy/autohttps -c traefik -- cat /etc/acme/acme.json

# these logs should contain "Register..."
kubectl logs deploy/autohttps -c traefik

# these logs should contain "Issued certificate"
kubectl logs deploy/pebble -c pebble
```

## Hub restarts

Have you seen the hub pod get a restart count > 0? JupyterHub 1.1.0 is typically
crashing after 20 seconds if it started up without the configurable proxy pod
available. This harmless error can be confirmed by doing a `kubectl logs

> deploy/hub --previous` if you spot a message about a timeout after ~20 000 ms in

Reviewer: ```suggestion
deploy/hub --previous` if you spot a message about a timeout after ~20 seconds in
```
